Mathlib/NumberTheory/NumberField/CanonicalEmbedding/FundamentalCone.lean

```diff
@@ -0,0 +1,242 @@
+  ext w
+  simp_rw [Pi.add_apply, logMap_apply]
+  rw [map_mul, map_mul, Real.log_mul, Real.log_mul hx hy, add_mul]
+  · ring
+  · exact mixedEmbedding.norm_ne_zero_iff.mp hx w
+  · exact mixedEmbedding.norm_ne_zero_iff.mp hy w
+
+theorem logMap_apply_of_norm_one {x : mixedSpace K} (hx : mixedEmbedding.norm x = 1)
+    {w : InfinitePlace K} (hw : w ≠ w₀) :
+    logMap x ⟨w, hw⟩ = mult w * Real.log (normAtPlace w x) := by
+  rw [logMap, hx, Real.log_one, zero_mul, sub_zero]
+
+theorem logMap_eq_logEmbedding (u : (𝓞 K)ˣ) :
+    logMap (mixedEmbedding K u) = logEmbedding K u := by
+  ext
+  rw [logMap_apply_of_norm_one (mixedEmbedding.norm_unit _), logEmbedding_component,
+    normAtPlace_apply]
+
+theorem logMap_unitSMul (u : (𝓞 K)ˣ) {x : mixedSpace K} (hx : mixedEmbedding.norm x ≠ 0) :
+    logMap (u • x) = logEmbedding K u + logMap x := by
+  rw [unitSMul_smul, logMap_mul (by rw [norm_unit]; norm_num) hx, logMap_eq_logEmbedding]
+
+theorem logMap_torsion_unitSMul (x : mixedSpace K) {ζ : (𝓞 K)ˣ} (hζ : ζ ∈ torsion K) :
+    logMap (ζ • x) = logMap x := by
+  ext
+  simp_rw [logMap_apply, unitSMul_smul, map_mul, norm_eq_norm, Units.norm, Rat.cast_one, one_mul,
+    normAtPlace_apply, (mem_torsion K).mp hζ, one_mul]
+
+theorem logMap_real (c : ℝ) :
+    logMap (c • (1 : mixedSpace K)) = 0 := by
+  ext
+  rw [logMap_apply, normAtPlace_smul, norm_smul, map_one, map_one, mul_one, mul_one, Real.log_pow,
+    mul_comm (finrank ℚ K : ℝ) _, mul_assoc, mul_inv_cancel₀ (Nat.cast_ne_zero.mpr
+    (ne_of_gt finrank_pos)), mul_one, sub_self, mul_zero, Pi.zero_apply]
+
+theorem logMap_real_smul {x : mixedSpace K} (hx : mixedEmbedding.norm x ≠ 0) {c : ℝ} (hc : c ≠ 0) :
+    logMap (c • x) = logMap x := by
+  have : mixedEmbedding.norm (c • (1 : mixedSpace K)) ≠ 0 := by
+    rw [norm_smul, map_one, mul_one]
+    exact pow_ne_zero _ (abs_ne_zero.mpr hc)
+  rw [← smul_one_mul, logMap_mul this hx, logMap_real, zero_add]
+
+theorem logMap_eq_of_normAtPlace_eq {x y : mixedSpace K}
+    (h : ∀ w, normAtPlace w x = normAtPlace w y) :
+    logMap x = logMap y := by
+  ext
+  simp_rw [logMap_apply, h, norm_eq_of_normAtPlace_eq h]
+
+end logMap
+
+noncomputable section
+
+open NumberField.Units NumberField.Units.dirichletUnitTheorem
+
+variable [NumberField K]
+
+open Classical in
+/-- The fundamental cone is a cone in the mixed space--that is a subset fixed by multiplication by
+a scalar, see `smul_mem_of_mem`--, that is also a fundamental domain for the action of `(𝓞 K)ˣ` up
+to roots of unity, see `exists_unitSMul_mem` and `torsion_unitSMul_mem_of_mem`. -/
+def fundamentalCone : Set (mixedSpace K) :=
+  logMap⁻¹' (ZSpan.fundamentalDomain ((basisUnitLattice K).ofZLatticeBasis ℝ _)) \
+      {x | mixedEmbedding.norm x = 0}
+
+namespace fundamentalCone
+
+variable {K}
+
+theorem norm_pos_of_mem {x : mixedSpace K} (hx : x ∈ fundamentalCone K) :
+    0 < mixedEmbedding.norm x :=
+  lt_iff_le_and_ne.mpr ⟨mixedEmbedding.norm_nonneg _, Ne.symm hx.2⟩
```

Comment:
```suggestion
  lt_of_le_of_ne (mixedEmbedding.norm_nonneg _) (Ne.symm hx.2)
```

Reply:
Another lemma I didn't know existed. Thanks for the pointer :)